core/state/statedb.go

```diff
@@ -1275,65 +1278,56 @@ func (s *StateDB) Commit(block uint64, deleteEmptyObjects bool) (common.Hash, er
 // - Reset transient storage (EIP-1153)
 func (s *StateDB) Prepare(rules params.Rules, sender, coinbase common.Address, dst *common.Address, precompiles []common.Address, list types.AccessList) {
 	if rules.IsBerlin {
+		if !rules.IsVerkle {
+			s.accessList = newAccessList()
+		} else {
+			s.accessList = NewAccessWitness()
+		}
 		// Clear out any leftover from previous executions
-		al := newAccessList()
-		s.accessList = al
+		al := s.accessList
 
-		al.AddAddress(sender)
+		al.AddAddress(sender, ALAllItems, AccessListRead)
 		if dst != nil {
-			al.AddAddress(*dst)
+			al.AddAddress(*dst, ALAllItems, AccessListRead)
 			// If it's a create-tx, the destination will be added inside evm.create
 		}
 		for _, addr := range precompiles {
-			al.AddAddress(addr)
+			al.AddAddress(addr, ALAllItems, AccessListRead)
 		}
 		for _, el := range list {
-			al.AddAddress(el.Address)
+			al.AddAddress(el.Address, 0, AccessListRead)
 			for _, key := range el.StorageKeys {
-				al.AddSlot(el.Address, key)
+				al.AddSlot(el.Address, key, AccessListRead)
 			}
 		}
 		if rules.IsShanghai { // EIP-3651: warm coinbase
-			al.AddAddress(coinbase)
+			al.AddAddress(coinbase, ALAllItems, AccessListRead)
 		}
 	}
 	// Reset transient storage at the beginning of transaction execution
 	s.transientStorage = newTransientStorage()
 }
 
 // AddAddressToAccessList adds the given address to the access list
-func (s *StateDB) AddAddressToAccessList(addr common.Address) {
-	if s.accessList.AddAddress(addr) {
+func (s *StateDB) AddAddressToAccessList(addr common.Address, mode ALAccountItem, isWrite ALAccessMode) uint64 {
+	gas := s.accessList.AddAddress(addr, mode, isWrite)
+	if gas > 0 {
 		s.journal.append(accessListAddAccountChange{&addr})
 	}
+	return gas
 }
 
 // AddSlotToAccessList adds the given (address, slot)-tuple to the access list
-func (s *StateDB) AddSlotToAccessList(addr common.Address, slot common.Hash) {
-	addrMod, slotMod := s.accessList.AddSlot(addr, slot)
-	if addrMod {
-		// In practice, this should not happen, since there is no way to enter the
-		// scope of 'address' without having the 'address' become already added
-		// to the access list (via call-variant, create, etc).
-		// Better safe than sorry, though
-		s.journal.append(accessListAddAccountChange{&addr})
-	}
```

Comment:
This change breaks tests. The comment itself says that this should not happen. My preference is to get rid of it and fix the tests: the reason is that in verkle, state and account are dissociated. It would be possible to link them, but it would be work done for a situation that should not occur in the first place.

Otherwise, I can add a `addrMod` return